File: map_ts/src/Exercices/Restaurant/04-product-calories.ts
import type {
  Ingredient,
  PreparedProduct,
  CalorieValue
} from "../../Restaurant/index.js";


const ingredients: Ingredient[] = [
  {
    id: "beef",
    name: "Beef",
    quantity: 3600,
    unit: "g",
    caloriesPerUnit: 2.5
  },
  {
    id: "chicken",
    name: "Chicken",
    quantity: 3600,
    unit: "g",
    caloriesPerUnit: 1.65
  },
  {
    id: "fish",
    name: "Fish",
    quantity: 3600,
    unit: "g",
    caloriesPerUnit: 1.2
  },
  {
    id: "potato",
    name: "Potato",
    quantity: 6000,
    unit: "g",
    caloriesPerUnit: 0.77
  },
  {
    id: "tomato",
    name: "Tomato",
    quantity: 6000,
    unit: "g",
    caloriesPerUnit: 0.18
  },
  {
    id: "salad",
    name: "Salad",
    quantity: 6000,
    unit: "g",
    caloriesPerUnit: 0.18
  },
  {
    id: "oignon",
    name: "Oignon",
    quantity: 6000,
    unit: "g",
    caloriesPerUnit: 0.32
  },
];


const products: PreparedProduct[] = [
  {
    id: "beef-plate",
    name: "Beef Plate",

    ingredients: [
      {
        ingredientId: "beef",
        quantity: 180
      },
      {
        ingredientId: "potato",
        quantity: 150
      }
    ],

    availableUnits: 20
  },
  {
    id: "chicken-plate",
    name: "Chicken Plate",

    ingredients: [
      {
        ingredientId: "chicken",
        quantity: 180
      },
      {
        ingredientId: "salad",
        quantity: 150
      }
    ],

    availableUnits: 30
  },
  {
    id: "fish-plate",
    name: "Fish Plate",

    ingredients: [
      {
        ingredientId: "fish",
        quantity: 180
      },
      {
        ingredientId: "oignon",
        quantity: 150
      }
    ],

    availableUnits: 50
  }

];


const calories: CalorieValue[] =
  products.map(product => {

    const ingredientCalories =
      product.ingredients.map(component => {

        // EXERCICE :
        //
        // rechercher l'ingrédient
        // puis calculer :
        let ingredient = ingredients.find(ingredient => ingredient.id === component.ingredientId);
        // component.quantity
        // *
        // ingredient.caloriesPerUnit
        if (ingredient) {
          return component.quantity * ingredient.caloriesPerUnit;
        };

        throw new Error("No ingredient found");
      });


    // EXERCICE :
    //
    // utiliser reduce() afin de transformer :
    //
    // [450, 115.5]
    //
    // en :
    //
    // 565.5

    const totalCalories = ingredientCalories.reduce((sum, calories) => sum + calories, 0);


    return {
      productId: product.id,
      calories: totalCalories
    };
  });


console.log(calories);

File: map_python/src/Exercises/Restaurant/04_product_calories.py
from functools import reduce

from src.Restaurant import CalorieValue, Ingredient, PreparedProduct, ProductIngredient


ingredients = [
    Ingredient("beef", "Beef", 3600, "g", 2.5),
    Ingredient("chicken", "Chicken", 3600, "g", 1.65),
    Ingredient("fish", "Fish", 3600, "g", 1.2),
    Ingredient("potato", "Potato", 6000, "g", 0.77),
    Ingredient("salad", "Salad", 6000, "g", 0.18),
    Ingredient("oignon", "Oignon", 6000, "g", 0.32),
]

products = [
    PreparedProduct(
        "beef-plate",
        "Beef Plate",
        [ProductIngredient("beef", 180), ProductIngredient("potato", 150)],
        20,
    ),
    PreparedProduct(
        "chicken-plate",
        "Chicken Plate",
        [ProductIngredient("chicken", 180), ProductIngredient("salad", 150)],
        20,
    ),
    PreparedProduct(
        "fish-plate",
        "Fish Plate",
        [ProductIngredient("fish", 180), ProductIngredient("oignon", 150)],
        20,
    ),
]


def ingredient_calories(component: ProductIngredient) -> float:
    ingredient = next(
        (item for item in ingredients if item.id == component.ingredient_id),
        None,
    )
    if ingredient is None:
        raise ValueError(f"Unknown ingredient: {component.ingredient_id}")
    return component.quantity * ingredient.calories_per_unit


calories = list(
    map(
        lambda product: CalorieValue(
            product_id=product.id,
            calories=reduce(
                lambda total, value: total + value,
                map(ingredient_calories, product.ingredients),
                0.0,
            ),
        ),
        products,
    )
)


if __name__ == "__main__":
    for value in calories:
        print(value)
